crates/stack/src/network.rs

```diff
@@ -960,6 +989,170 @@ mod tests {
+            UnboundedReceiverStream::new(net3.ingress_receiver().unwrap()),
+        );
+
+        let _consume3 = spawn_local(rx.for_each(|e| async move { println!("consumer 3: {e:?}") }));
+
+        let conn1 = net2.connect((ip1, 1), Duration::from_secs(3));
+        let conn2 = net3.connect((ip1, 1), Duration::from_secs(3));
+
+        let (f1, f2) = futures::future::join(conn1, conn2).await;
+
+        f1.expect("Connection failed!");
+        f2.expect("Connection failed!");
+
+        Ok(())
+    }
+
+    /// Establish given number of connections between single client and server
+    async fn establish_multiple_conn(
+        medium: Medium,
+        total: usize,
+        chunk_size: usize,
+        conn_num: u16,
+    ) -> anyhow::Result<()> {
+        const MTU: usize = 65535;
+
+        println!(">> exchanging {} B in {} B chunks", total, chunk_size);
+
+        let ip1 = Ipv4Address::new(10, 0, 0, 1);
+        let ip2 = Ipv4Address::new(10, 0, 0, 2);
+
+        let config = StackConfig {
+            max_transmission_unit: MTU,
+            ..Default::default()
+        };
+
+        let net1 = new_network(medium, ip1.into(), config.clone());
+        let net2 = new_network(medium, ip2.into(), config.clone());
+
+        net1.spawn_local();
+        net2.spawn_local();
+
+        net1.bind(Protocol::Tcp, (ip1, 1))?;
+        net2.bind(Protocol::Tcp, (ip2, 1))?;
+
+        // net 1
+        // inject egress packets from net 2 into net 1 rx buffer
+        net_inject(
+            UnboundedReceiverStream::new(net2.egress_receiver().unwrap())
+                .map(|e| e.payload.into_vec()),
+            net1.clone(),
+        );
+
+        // process net 1 events
+        let (tx, rx) = mpsc::channel(1);
+        net_receive(
+            tx,
+            UnboundedReceiverStream::new(net1.ingress_receiver().unwrap()),
+        );
+
+        let _consume1 = spawn_local(rx.for_each(|e| async move { println!("consumer 1: {e:?}") }));
+
+        // net 2
+        // inject egress packets from net 1 into net 2 rx buffer
+        net_inject(
+            UnboundedReceiverStream::new(net1.egress_receiver().unwrap())
+                .map(|e| e.payload.into_vec()),
+            net2.clone(),
+        );
+
+        // process net 2 events
+        let (tx, rx) = mpsc::channel(1);
+        net_receive(
+            tx,
+            UnboundedReceiverStream::new(net2.ingress_receiver().unwrap()),
+        );
+
+        let _consume2 = spawn_local(rx.for_each(|e| async move { println!("consumer 2: {e:?}") }));
+
+        for i in 1..=conn_num {
+            let conn = net2.connect((ip1, 1), Duration::from_secs(3)).await;
+            match conn {
+                Ok(_) => println!("Connection({i}) successful"),
+                Err(_) => {
+                    if i != u16::MAX {
+                        println!("Connection failed!")
+                    } else {
+                        return Ok(());
+                    }
+                }
```

Comment:
Is that really supposed to be `Err(_)`? I assumed that only one specific kind of error would get a pass on `i == u16::MAX`, but if establishing connection fails for a different reason it should also be logged instead of just returning `Ok(())`.

I don't remember what variants (if any) this error type has, so my complaint might be irrelevant.

Reply:
So indeed it has not been implemented correctly. The test should fail every time when connection fails. The only exception when it should pass on error is when `i == u16::MAX` and error asserts to be `Error::Other("no ports available".into())`